File: sorting_algorithms/merge_sort/scala/merge_sort.scala
object merge_sort extends App {
    def mergeSort(arr2: Array[Int], start: Int, end: Int, arr: Array[Int]) : Unit = {
        if (end - start > 1) {
            val mid : Int = ((start + end) / 2).toInt
            mergeSort(arr, start, mid, arr2)
            mergeSort(arr, mid, end, arr2)
            merge(arr2, start, mid, end, arr)
        }
    }

    def merge(arr2: Array[Int], start: Int, mid: Int, end: Int, arr: Array[Int]) : Unit = {
        var i : Int = start
        var j : Int = mid

        (start until end).foreach(k => {
            if (i < mid && (j >= end || arr(i) <= arr(j))) {
                arr2(k) = arr(i)
                i += 1
            } else {
                arr2(k) = arr(j)
                j += 1
            }
        })
    }

    var arr : Array[Int] = Array(0, 4, 1, 3, 2, 16, 9, 10, 14, 8, 7)
    var arr2 : Array[Int] = arr.clone()
    mergeSort(arr2, 0, arr.length, arr)
    println(arr2.mkString(" "))
}

File: sorting_algorithms/merge_sort/java/merge_sort.java
package sorting_algorithms.merge_sort.java;

public class merge_sort {
    public static void merge(int[] A, int start, int mid, int end, int[] B) {
        int i = start, j = mid;
        for (int k = start; k < end; k++) {
            if (i < mid && (j >= end || A[i] <= A[j])) {
                B[k] = A[i];
                i = i + 1;
            } else {
                B[k] = A[j];
                j = j + 1;
            }
        }
    }

    public static void mergeSort(int[] B, int start, int end, int[] A) {
        if (end - start < 2) {
            return;
        }
        int mid = (start + end) / 2;
        mergeSort(A, start, mid, B);
        mergeSort(A, mid, end, B);
        merge(B, start, mid, end, A);
    }

    public static void main(String[] args) {
        int[] A = { 5, 4, 3, 2, 1 };
        int[] B = A.clone();
        mergeSort(B, 0, A.length, A);
        for (int i = 0; i < A.length; i++) {
            System.out.print(A[i] + " ");
        }
    }
}